Improving typing of `xr.Dataset.__getitem__`

First, I'd like the thank the xarray dev's for adding type hints to this library, not many libraries have this feature!

That said, the indexing notation of `xr.Dataset` does not currently play well wit mypy since it returns a Union type. This results in a lot of mypy errors like this: 
```
workflows/fine_res_budget/budget/budgets.py:284: error: Argument 6 to "compute_recoarsened_budget_field" has incompatible type "Union[DataArray, Dataset]"; expected "DataArray"
workflows/fine_res_budget/budget/budgets.py:285: error: Argument 1 to "storage" has incompatible type "Union[DataArray, Dataset]"; expected "DataArray"
workflows/fine_res_budget/budget/budgets.py:286: error: Argument "unresolved_flux" to "compute_recoarsened_budget_field" has incompatible type "Union[DataArray, Dataset]"; expected "DataArray"
workflows/fine_res_budget/budget/budgets.py:287: error: Argument "saturation_adjustment" to "compute_recoarsened_budget_field" has incompatible type "Union[DataArray, Dataset]"; expected "DataArray"
```

#### MCVE Code Sample
<!-- In order for the maintainers to efficiently understand and prioritize issues, we ask you post a "Minimal, Complete and Verifiable Example" (MCVE): http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports -->

```
def func(ds: xr.Dataset):
    pass

dataset: xr.Dataset = ...

# error:
# this line will give type error because mypy doesn't know 
# if ds[['a', 'b]] is Dataset or a DataArray
func(ds[['a', 'b']])
```

#### Expected Output

Mypy should be able to infer that `ds[['a', b']]` is a Dataset, and that `ds['a']` is a DataArray.

#### Problem Description

This requires any routine with type hints that consume an output of `xr.Dataset.__getitem__` to require a `Union[DataArray, Dataset]` even if it really intends to  be used with either `DataArray` or `DataArray`. Because `ds[something]` is a ubiquitous syntax,  this behavior accounts for approximately 50% of mypy errors in my xarray heavy code.


#### Versions

<details><summary>Output of <tt>xr.show_versions()</tt></summary>


In [1]: import xarray as xr                                                                                                                                                                          
xr.
In [2]: xr.show_versions()                                                                                                                                                                           

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.7 (default, May  7 2020, 21:25:33) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 5.3.0-1020-gcp
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.4
libnetcdf: 4.7.3

xarray: 0.15.1
pandas: 1.0.1
numpy: 1.18.1
scipy: 1.4.1
netCDF4: 1.5.3
pydap: None
h5netcdf: 0.8.0
h5py: 2.10.0
Nio: None
zarr: 2.4.0
cftime: 1.1.2
nc_time_axis: 1.2.0
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2.17.2
distributed: 2.17.0
matplotlib: 3.1.3
cartopy: 0.17.0
seaborn: 0.10.1
numbagg: None
setuptools: 46.4.0.post20200518
pip: 20.0.2
conda: 4.8.3
pytest: 5.4.2
IPython: 7.13.0
sphinx: None
</details>


# Potential solution

I think we can fix this with [typing.overload](https://docs.python.org/3/library/typing.html#typing.overload). I am not too familiar with that librariy, but I think something like the following might work:

```
from typing import overload

class Dataset
    @overload
    def __getitem__(self, key: Hashable) -> DataArray: ...
    
     @overload
    def __getitem__(self, key: List[Hashable]) -> "Dataset": ...

     # actual implementation
    def __getitem__
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -16,7 +16,7 @@ repos:
     hooks:
       - id: flake8
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.761  # Must match ci/requirements/*.yml
+    rev: v0.780  # Must match ci/requirements/*.yml
     hooks:
       - id: mypy
   # run this occasionally, ref discussion https://github.com/pydata/xarray/pull/3194

--- a/ci/requirements/py38.yml
+++ b/ci/requirements/py38.yml
@@ -22,7 +22,7 @@ dependencies:
   - isort
   - lxml    # Optional dep of pydap
   - matplotlib
-  - mypy=0.761  # Must match .pre-commit-config.yaml
+  - mypy=0.780  # Must match .pre-commit-config.yaml
   - nc-time-axis
   - netcdf4
   - numba

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1243,7 +1243,7 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         return _LocIndexer(self)
 
     @overload  # noqa: F811
-    def __getitem__(self, key: Hashable) -> DataArray:
+    def __getitem__(self, key: Hashable) -> "DataArray":
         ...
 
     @overload  # noqa: F811

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1242,15 +1242,15 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         """
         return _LocIndexer(self)
 
-    @overload
+    @overload  # noqa: F811
     def __getitem__(self, key: Hashable) -> DataArray:
         ...
 
-    @overload
+    @overload  # noqa: F811
     def __getitem__(self, key: Any) -> "Union[DataArray, Dataset]":
         ...
 
-    def __getitem__(self, key):
+    def __getitem__(self, key):  # noqa: F811
         """Access variables or coordinates this dataset as a
         :py:class:`~xarray.DataArray`.
 

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1241,13 +1241,18 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         and only when the key is a dict of the form {dim: labels}.
         """
         return _LocIndexer(self)
-    
 
     @overload
-    def __getitem__(self, key: Hashable) -> DataArray: ...
+    def __getitem__(self, key: Hashable) -> DataArray:
+        ...
 
     @overload
-    def __getitem__(self, key: Any) -> Dataset: ...
+    def __getitem__(self, key: Mapping) -> "Dataset":
+        ...
+
+    @overload
+    def __getitem__(self, key: List) -> "Dataset":
+        ...
 
     def __getitem__(self, key):
         """Access variables or coordinates this dataset as a

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1242,21 +1242,25 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         """
         return _LocIndexer(self)
 
-    @overload  # noqa: F811
-    def __getitem__(self, key: Hashable) -> "DataArray":
+    # FIXME https://github.com/python/mypy/issues/7328
+    @overload
+    def __getitem__(self, key: Mapping) -> "Dataset":  # type: ignore
         ...
 
-    @overload  # noqa: F811
-    def __getitem__(self, key: Any) -> "Union[DataArray, Dataset]":
+    @overload
+    def __getitem__(self, key: Hashable) -> "DataArray":  # type: ignore
         ...
 
-    def __getitem__(self, key):  # noqa: F811
+    @overload
+    def __getitem__(self, key: Any) -> "Dataset":
+        ...
+
+    def __getitem__(self, key):
         """Access variables or coordinates this dataset as a
         :py:class:`~xarray.DataArray`.
 
         Indexing with a list of names will return a new ``Dataset`` object.
         """
-        # TODO(shoyer): type this properly: https://github.com/python/mypy/issues/7328
         if utils.is_dict_like(key):
             return self.isel(**cast(Mapping, key))
 

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -1247,11 +1247,7 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         ...
 
     @overload
-    def __getitem__(self, key: Mapping) -> "Dataset":
-        ...
-
-    @overload
-    def __getitem__(self, key: List) -> "Dataset":
+    def __getitem__(self, key: Any) -> "Union[DataArray, Dataset]":
         ...
 
     def __getitem__(self, key):

--- a/xarray/core/dataset.py
+++ b/xarray/core/dataset.py
@@ -27,6 +27,7 @@ from typing import (
     TypeVar,
     Union,
     cast,
+    overload,
 )
 
 import numpy as np
@@ -1240,8 +1241,15 @@ class Dataset(Mapping, ImplementsDatasetReduce, DataWithCoords):
         and only when the key is a dict of the form {dim: labels}.
         """
         return _LocIndexer(self)
+    
 
-    def __getitem__(self, key: Any) -> "Union[DataArray, Dataset]":
+    @overload
+    def __getitem__(self, key: Hashable) -> DataArray: ...
+
+    @overload
+    def __getitem__(self, key: Any) -> Dataset: ...
+
+    def __getitem__(self, key):
         """Access variables or coordinates this dataset as a
         :py:class:`~xarray.DataArray`.
 

--- a/xarray/core/weighted.py
+++ b/xarray/core/weighted.py
@@ -72,11 +72,11 @@ class Weighted:
     def __init__(self, obj: "DataArray", weights: "DataArray") -> None:
         ...
 
-    @overload  # noqa: F811
-    def __init__(self, obj: "Dataset", weights: "DataArray") -> None:  # noqa: F811
+    @overload
+    def __init__(self, obj: "Dataset", weights: "DataArray") -> None:
         ...
 
-    def __init__(self, obj, weights):  # noqa: F811
+    def __init__(self, obj, weights):
         """
         Create a Weighted object
 

